cassiopeia/solver/UPGMASolver.py

```diff
@@ -0,0 +1,213 @@
+    dissimilarity between samples. After joining nodes, the dissimilarities
+    are updated by averaging the distances of elements in the new cluster 
+    with each existing node. Produces a rooted tree that is assumed to be
+    ultrametric.
+
+
+    Args:
+        dissimilarity_function: A function by which to compute the dissimilarity
+            map. Optional if a dissimilarity map is already provided.
+        prior_transformation: Function to use when transforming priors into
+            weights. Supports the following transformations:
+                "negative_log": Transforms each probability by the negative
+                    log (default)
+                "inverse": Transforms each probability p by taking 1/p
+                "square_root_inverse": Transforms each probability by the
+                    the square root of 1/p
+
+    """
+
+    def __init__(
+        self,
+        dissimilarity_function: Optional[
+            Callable[
+                [np.array, np.array, int, Dict[int, Dict[int, float]]], float
+            ]
+        ] = None,
+        prior_transformation: str = "negative_log",
+    ):
+
+        super().__init__(
+            dissimilarity_function=dissimilarity_function,
+            add_root=True,
+            prior_transformation=prior_transformation,
+        )
+
+
+    def root_tree(self, tree: nx.Graph, root_sample: str, remaining_samples: List[str]):
+        """Roots the tree.
+
+        Adds the root at the top of the UPGMA reconstructed tree. By the 
+        ultrametric assumption, the root is placed as the parent to the last 
+        two unjoined nodes.
+
+        Args:
+            tree: Networkx object representing the tree topology
+            root_sample: Ignored in this case, the root is known in this case
+            remaining_samples: The last two unjoined nodes in the tree
+
+        Returns:
+            A rooted tree.
+        """
+
+        tree.add_node("root")
+        tree.add_edges_from([("root", remaining_samples[0]), ("root", remaining_samples[1])])
+        
+        rooted_tree = nx.DiGraph()
+        for e in nx.dfs_edges(tree, source="root"):
+            rooted_tree.add_edge(e[0], e[1])
+
+        return rooted_tree
+
+
+    def find_cherry(self, dissimilarity_matrix: np.array) -> Tuple[int, int]:
+        """Finds a pair of samples to join into a cherry.
+
+        Finds the pair of samples with the minimum dissimilarity by finding the
+        minimum value in the provided dissimilarity matrix
+
+        Args:
+            dissimilarity_matrix: A sample x sample dissimilarity matrix
+
+        Returns:
+            A tuple of integers representing rows in the dissimilarity matrix
+                to join.
+        """
+
+        dissimilarity_matrix = dissimilarity_matrix.astype(float)
+        np.fill_diagonal(dissimilarity_matrix, np.inf)
+
+        return np.unravel_index(np.argmin(dissimilarity_matrix, axis=None), dissimilarity_matrix.shape)
```

Comment:
Is this better/faster than the way I find the `argmin` in `NeighborJoiningSolver`? If so, let's change the way I find the indices in `NeighborJoiningSolver`